### PR TITLE
FM2-174: Improve search for DiagnosticReport resource

### PR DESCRIPTION
**Pull request by varung-31**
## Description of what I changed

Added search parameters to DiagnosticReport resource like:
1. encounterReference
2. patientReference
3. issueDate
4. effectiveDate
5. code

Also, added sorting of the results by effectiveDate and issueDate.

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-174

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**.

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by varung-31** [comment]
@ibacher  Could you take a look? The build is failing here but works perfectly on my computer. Looks like some commit in the openmrs-api is failing it.

**Comment by ibacher** [comment]
@varung-31 Looks like a failure on the Travis end. In particular, the code to run on Java 8 was missing so it instead ran on Java 11.

File: omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java [hunk 1]
```diff
@@ -69,4 +79,18 @@ public MethodOutcome updateDiagnosticReport(@IdParam IdType id, @ResourceParam D
...
+	
+	@Search
+	public Bundle searchForDiagnosticReports(
+	        @OptionalParam(name = DiagnosticReport.SP_ENCOUNTER, chainWhitelist = {
+	                "" }, targetTypes = Encounter.class) ReferenceAndListParam encounterReference,
+	        @OptionalParam(name = DiagnosticReport.SP_PATIENT, chainWhitelist = { "", Patient.SP_IDENTIFIER,
+	                Patient.SP_GIVEN, Patient.SP_FAMILY,
+	                Patient.SP_NAME }, targetTypes = Patient.class) ReferenceAndListParam patientReference,
+	        @OptionalParam(name = DiagnosticReport.SP_ISSUED) DateRangeParam issueDate,
+	        @OptionalParam(name = DiagnosticReport.SP_DATE) DateRangeParam effectiveDate,
```
Comment (ibacher) [inline omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java:91]: Let's not try to map an effectiveDate. We don't really have anything for which that's semantically meaningful.

File: api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml [hunk 2]
```diff
@@ -18,5 +18,5 @@
...
-	<obs obs_id="19" person_id="7" concept_id="5085" encounter_id="3" obs_datetime="2008-07-01 00:00:00.0" location_id="1" value_numeric="[NULL]" comments="" creator="1" date_created="2008-08-18 14:09:35.0" voided="false" value_coded="[NULL]" value_coded_name_id="[NULL]" value_complex="[NULL]" value_text="LAB" value_datetime="[NULL]" value_drug="[NULL]" uuid="dc386962-1c42-49ea-bed2-97650c66f742"/>
+	<obs obs_id="19" person_id="8" concept_id="5085" encounter_id="3" obs_datetime="2008-07-01 00:00:00.0" location_id="1" value_numeric="[NULL]" comments="" creator="1" date_created="2008-08-18 14:09:35.0" voided="false" value_coded="[NULL]" value_coded_name_id="[NULL]" value_complex="[NULL]" value_text="LAB" value_datetime="[NULL]" value_drug="[NULL]" uuid="dc386962-1c42-49ea-bed2-97650c66f742"/>
```
Comment (ibacher) [inline api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml:21]: Why did you change the person associated with this obs?

File: omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java (hunk 1 above)
Comment (varung-31) [reply omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java:91]: But we are doing that for observation as well.

File: api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml (hunk 2 above)
Comment (varung-31) [reply api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml:21]: It becomes difficult to test if all the data is similar. But I could revert if you want.

File: omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java (hunk 1 above)
Comment (ibacher) [reply omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java:91]: The meaning is a bit different for a DiagnosticReport versus a standard observation. One of the reasons I want to get them off of this weird obs group formatting.

File: api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml (hunk 2 above)
Comment (ibacher) [reply api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml:21]: That's reasonable. Leave it as is.

File: omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java (hunk 1 above)
Comment (varung-31) [reply omod/src/main/java/org/openmrs/module/fhir2/providers/DiagnosticReportFhirResourceProvider.java:91]: Okay, i'll do that.

File: api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml (hunk 2 above)
Comment (jecihjoy) [reply api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml:21]: @varung-31 another option would be to add more test data if the existing data doesn't suite your test cases

**Comment by jecihjoy** [comment]
> @varung-31 Looks like a failure on the Travis end. In particular, the code to run on Java 8 was missing so it instead ran on Java 11.

@ibacher i've been getting the same error whole day and actually spent sometime trying out suggestions from google which none worked, it was an issue with powerMockito, is there something that needs to be fixed in travis? we're already using jdk8, what would be the reason travis would lack the code to run java 8?

**Comment by ibacher** [comment]
@jecihjoy Travis build environments have multiple Java versions installed an run a script to configure the correct default version. That script failed to run correctly (I don't know why) so it ran under the system default, which appeared to be Java 11. The best I can say is try re-running the build. That seemed to fix things here.

File: api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml (hunk 2 above)
Comment (ibacher) [reply api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml:21]: @varung-31 I agree with @jecihjoy's suggestion here, that it would be better to add some new data rather than deleting existing data.

Comment (varung-31) [reply api/src/test/resources/org/openmrs/module/fhir2/api/dao/impl/FhirDiagnosticReportDaoImplTest_initial_data.xml]: Done

**Comment by ibacher** [comment]
@varung-31 Thanks for this!
